ext/sockets/sockets.c

```diff
@@ -1416,6 +1435,57 @@ PHP_FUNCTION(socket_bind)
 }
 /* }}} */
 
+#ifdef AF_PACKET
+#define ETH_SUB_CHECKLENGTH(a, lyr)						\
+	do {									\
+		if ((char *)ipdata + sizeof(a) < ZSTR_VAL(recv_buf) + slen) {	\
+			return FAILURE;						\
+		}								\
+	} while (0)
+
+static zend_result php_socket_get_chunk(zend_string *dst, const zend_string *src, size_t offset, size_t len) {
+	if (UNEXPECTED((offset > SIZE_MAX - len) || offset + len > ZSTR_LEN(src))) {
+		return FAILURE;
+	}
+
+	memcpy(ZSTR_VAL(dst), ZSTR_VAL(src) + offset, len + 1);
+	ZSTR_LEN(dst) = len;
+	return SUCCESS;
+}
+
+static zend_result php_socket_afpacket_add_tcp(unsigned char *ipdata, struct sockaddr_ll sll, char *ifrname, zend_string *recv_buf,
+					       size_t slen, zval *szpayload,  zval *zpayload, zval *obj, zval *data, zval *addr, zval *index) {
+	struct tcphdr tcp;
+	if ((char *)ipdata + sizeof(tcp) < ZSTR_VAL(recv_buf) + slen)
+		return FAILURE;
```

Comment:
I don't understand this check. Is it necessary, when the length of `ipdata` is already checked by `php_socket_get_chunk()`? Same remark for `php_socket_afpacket_add_udp()`.